File: scripts/install-cargo-audit.ts
import { createRequire } from 'module';
import { fileURLToPath } from 'url';
import { dirname } from 'path';
const require = createRequire(import.meta.url);
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const https = require('https');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const url = 'https://github.com/rustsec/rustsec/releases/download/cargo-audit/v0.20.1/cargo-audit-x86_64-pc-windows-msvc.zip';
const tempZip = path.join(process.env.TEMP, 'cargo-audit.zip');
const destDir = path.join(process.env.USERPROFILE, '.cargo', 'bin');

console.log('Downloading cargo-audit...');
const file = fs.createWriteStream(tempZip);
https
  .get(url, (response) => {
    if (response.statusCode === 301 || response.statusCode === 302) {
      https.get(response.headers.location, (res) => {
        res.pipe(file);
        file.on('finish', () => {
          file.close(() => {
            console.log('Download complete. Extracting...');
            try {
              execSync(`powershell -Command "Expand-Archive -Path '${tempZip}' -DestinationPath '${process.env.TEMP}\\cargo-audit-ext' -Force"`);
              const exePath = path.join(process.env.TEMP, 'cargo-audit-ext', 'cargo-audit-x86_64-pc-windows-msvc', 'cargo-audit.exe');
              if (!fs.existsSync(destDir)) fs.mkdirSync(destDir, { recursive: true });
              fs.copyFileSync(exePath, path.join(destDir, 'cargo-audit.exe'));
              console.log('cargo-audit installed successfully to .cargo/bin!');
            } catch (e) {
              console.error('Failed to extract:', e.message);
            }
          });
        });
      });
    }
  })
  .on('error', (err) => {
    fs.unlink(tempZip, () => {});
    console.error('Error downloading:', err.message);
  });

File: scripts/install-cargo-audit.py
import urllib.request
import json
import zipfile
import os
import shutil

print('Fetching release info...')
api_url = 'https://api.github.com/repos/rustsec/rustsec/releases/latest'
req = urllib.request.Request(api_url, headers={'User-Agent': 'Mozilla/5.0'})
with urllib.request.urlopen(req) as response:
    data = json.loads(response.read().decode())

download_url = None
for asset in data['assets']:
    if 'cargo-audit' in asset['name'] and 'x86_64-pc-windows-msvc' in asset['name']:
        download_url = asset['browser_download_url']
        break

if not download_url:
    print('Could not find Windows binary asset.')
    exit(1)

print(f'Downloading from {download_url}...')
urllib.request.urlretrieve(download_url, 'cargo-audit.zip')

print('Extracting...')
with zipfile.ZipFile('cargo-audit.zip', 'r') as zip_ref:
    zip_ref.extractall('cargo-audit')

dest = os.path.join(os.environ['USERPROFILE'], '.cargo', 'bin')
os.makedirs(dest, exist_ok=True)
for root, _, files in os.walk('cargo-audit'):
    for file in files:
        if file == 'cargo-audit.exe':
            shutil.copy(os.path.join(root, file), dest)
            print('Installed!')
            exit(0)
print('Failed to find cargo-audit.exe in zip.')
